src/workqueue/workqueue.c

```diff
@@ -0,0 +1,849 @@
+// TODO: This may be moved "down" to user-level scheduler
+//       if it is not easy to maintain from the kext.
+static volatile int32_t runnable_workers;
+
+// The desired minimum number of runnable worker threads
+static int32_t target_runnable_workers;
+
+// The desired maximum number of runnable worker threads
+static int32_t target_max_runnable_workers;
+
+// Limit on the total number of worker threads that can be created.
+static int32_t max_total_workers;
+
+#if DISPATCH_ENABLE_PWQ_KEXT
+// Are we using user-level or kext based management?
+static bool kext_active;
+#endif
+
+// Called on startup of the workqueue system
+static void _dispatch_workqueue_management_init(void);
+
+// Called by worker thread when it starts executing
+static int _dispatch_worker_register(void);
+
+// Called by worker thread just before it exits
+static void _dispatch_worker_unregister(void);
+
+// Called when new work is added to the system
+static void _dispatch_work_added(void);
+
+// Called by a worker thread when either it has found no work to do
+// or it has been command to wait by the management subsystem.
+// This call is expected to block until the management subsystem
+// thinks there may be work for the thread to do.
+static void _dispatch_worker_wait(bool was_ordered_to_wait);
+
+// Called by the management system when it wants to inject a
+// workitem that will case a worker thread to wait
+static void _dispatch_worker_should_wait(void);
+
+// Called by the management system when it wants to inject a
+// workitem that will case a worker thread to terminate
+static void _dispatch_worker_should_exit(void);
+
+// Called by the management system when it wants to
+// create an additional worker thread.
+static void _dispatch_worker_should_start(void);
+
+#if DISPATCH_ENABLE_PWQ_KEXT
+//////////////////////////////////////////////////////////////////
+// versions of above functions required for the kernel extension
+// temporary to isolate kext code to separate file.
+//////////////////////////////////////////////////////////////////
+
+int _dispatch_workqueue_management_init_kext(void);
+void _dispatch_worker_register_kext(void);
+void _dispatch_worker_unregister_kext(void);
+void _dispatch_work_added_kext(void);
+void _dispatch_worker_wait_kext(void);
+void* dispatch_thread_monitor_main_kext(void *context DISPATCH_UNUSED);
+
+#endif // DISPATCH_ENABLE_PWQ_KEXT
+
+/////////////////////////////////
+// Workqueue internal data structures and functions
+/////////////////////////////////
+
+// Debug messages for threadpool operations
+static bool threadpool_verbose;
+#define verbose_printf(fmt,...) do {								\
+	if (unlikely(threadpool_verbose))								\
+		fprintf(stderr, "libdispatch_threadpool: "fmt"\n", ##__VA_ARGS__); \
+} while (0)
+
+struct work {
+	STAILQ_ENTRY(work) item_entry;
+	void (*func)(void *);
+	void *func_arg;
+};
+typedef struct work * dispatch_workitem_t;
+
+typedef struct _dispatch_workqueue {
```

Comment:
given my naming remark, this should actually likely actually be `dispatch_workq_item_t`